plonky2/src/recursion/cyclic_recursion.rs

```diff
@@ -0,0 +1,490 @@
+#![allow(clippy::int_plus_one)] // Makes more sense for some inequalities below.
+use anyhow::{ensure, Result};
+use itertools::Itertools;
+use plonky2_field::extension::Extendable;
+
+use crate::gates::noop::NoopGate;
+use crate::hash::hash_types::{HashOut, HashOutTarget, MerkleCapTarget, RichField};
+use crate::hash::merkle_tree::MerkleCap;
+use crate::iop::target::{BoolTarget, Target};
+use crate::iop::witness::{PartialWitness, Witness};
+use crate::plonk::circuit_builder::CircuitBuilder;
+use crate::plonk::circuit_data::{
+    CircuitData, CommonCircuitData, VerifierCircuitTarget, VerifierOnlyCircuitData,
+};
+use crate::plonk::config::Hasher;
+use crate::plonk::config::{AlgebraicHasher, GenericConfig};
+use crate::plonk::proof::{ProofWithPublicInputs, ProofWithPublicInputsTarget};
+use crate::recursion::conditional_recursive_verifier::dummy_proof;
+
+pub struct CyclicRecursionData<
+    'a,
+    F: RichField + Extendable<D>,
+    C: GenericConfig<D, F = F>,
+    const D: usize,
+> {
+    proof: &'a Option<ProofWithPublicInputs<F, C, D>>,
+    verifier_data: &'a VerifierOnlyCircuitData<C, D>,
+    common_data: &'a CommonCircuitData<F, C, D>,
+}
+
+pub struct CyclicRecursionTarget<const D: usize> {
+    pub proof: ProofWithPublicInputsTarget<D>,
+    pub verifier_data: VerifierCircuitTarget,
+    pub dummy_proof: ProofWithPublicInputsTarget<D>,
+    pub dummy_verifier_data: VerifierCircuitTarget,
+    pub base_case: BoolTarget,
+}
+
+pub struct CyclicPublicInputs<
+    F: RichField + Extendable<D>,
+    C: GenericConfig<D, F = F>,
+    const D: usize,
+> {
+    pub circuit_digest: HashOut<F>,
+    pub constants_sigmas_cap: MerkleCap<F, C::Hasher>,
+    pub base_case: bool,
```

Comment:
Thinking about this more, I'm not sure it should be a public input? I think would callers typically have some logic like "either the proof must perform cyclic recursion, or if it doesn't, then X", where X could be something like
- for a hash chain, "the preimage must be zero"
- for a chain of block proofs, "the initial state trie must be empty"
- for our aggregation tree, "a (non-dummy) transaction proof must be recursive verified"

I think it may be best to have `cyclic_recursion` just take any `condition: BoolTarget` (like `conditionally_verify_proof` does), and have the caller handle the rest.

Reply:
I don't think we necessarily need to know if the previous proof was a base case? It seems like the first proof can enforce a base-case condition on its own, e.g. for a hash chain we'd have logic like "if there's no inner proof, then our preimage must be zero".

Reply:
Hm but then how do you ensure in general that a base case is not following a recursion case? This might not be strictly necessary for certain use cases, but I think it could be useful in general.

Reply:
As per our discussion, I removed `base_case` from the public inputs. I also removed the `CyclicPublicInputs{,Target}` structs altogether, since they're now just equivalent to `VerifierOnlyCircuitData{,Target}`.